File: backend/app/schemas/feed_planner.py
from datetime import datetime

from pydantic import BaseModel, Field


class FeedGridItem(BaseModel):
    """A single item in the Instagram feed grid preview."""

    post_id: str
    thumbnail_url: str | None = None
    caption_preview: str = ""
    status: str  # published, scheduled, draft
    scheduled_time: datetime | None = None
    grid_position: int = Field(ge=0, description="Zero-based position in the grid")
    row: int = Field(ge=0, description="Zero-based row index in the 3-column grid")
    col: int = Field(ge=0, description="Zero-based column index (0, 1, or 2)")

    model_config = {"from_attributes": True}


class FeedGridResponse(BaseModel):
    """Response containing the full Instagram feed grid preview."""

    account_id: str
    platform: str = "instagram"
    items: list[FeedGridItem]
    total_published: int
    total_scheduled: int


class ReorderRequest(BaseModel):
    """Request to reorder scheduled posts in the feed grid."""

    post_ids: list[str] = Field(
        ...,
        min_length=1,
        description="Ordered list of scheduled post IDs in the desired display order",
    )


class ReorderResponse(BaseModel):
    """Response after reordering scheduled posts."""

    success: bool
    message: str

File: frontend/src/api/feedPlanner.ts
import { api } from './client';

export interface FeedGridItem {
  post_id: string;
  thumbnail_url: string | null;
  caption_preview: string;
  status: string;
  scheduled_time: string | null;
  grid_position: number;
  row: number;
  col: number;
}

export interface FeedGridResponse {
  account_id: string;
  platform: string;
  items: FeedGridItem[];
  total_published: number;
  total_scheduled: number;
}

export const getFeedPreview = (accountId: string, limit?: number) =>
  api.get<FeedGridResponse>(`/feed-planner/${accountId}`, { params: { limit } }).then(r => r.data);

export const reorderFeedPosts = (accountId: string, postIds: string[]) =>
  api.put(`/feed-planner/${accountId}/reorder`, { post_ids: postIds }).then(r => r.data);
